File: Master_the_Key_ROS_Concepts/Lec1_Create_a_Cakin_Workspace/catkin_ws/src/my_number_counter/scripts/oop_number_counter.py
#!/usr/bin/env python3

import rospy
from std_msgs.msg import Int64
from std_srvs.srv import SetBool

class NumberCounter:
    def __init__(self):
        # counter=0
        # pub=None
        self.counter = 0

        #subscriber
        self.number_subscribe = rospy.Subscriber("/number", Int64, self.callback_receive_number_data)
        
        #publisher
        self.pub = rospy.Publisher("/number_count", Int64, queue_size=10)

        #create Service server
        self.reset_service = rospy.Service("/reset_number_count", SetBool, self.reset_count)
        
    def callback_receive_number_data(self, msg):
        # global counter
        rospy.loginfo("Message Received")
        rospy.loginfo(msg)#msg.data also fine
        # counter += msg.data
        self.counter += msg.data
        new_msg = Int64()
        # new_msg.data = counter
        new_msg.data = self.counter
        # pub.publish(new_msg)
        self.pub.publish(new_msg)

    def reset_count(self, req):
        if req.data:
            # global counter
            self.counter = 0
            rospy.loginfo("Counter Reset!")
            # counter = 0
            return True, "Counter Reset Success"

        else:
            return False, "Counter Reset False"

if __name__ == '__main__':
    rospy.init_node('number_counter')
    NumberCounter()
    # #subscriber
    # sub = rospy.Subscriber("/number", Int64, callback_receive_number_data)
    
    # #publisher
    # pub = rospy.Publisher("/number_count", Int64, queue_size=10)

    # #create Service server
    # service = rospy.Service("/reset_number_count", SetBool, reset_count)

    rospy.spin()

File: Master_the_Key_ROS_Concepts/Lec1_Create_a_Cakin_Workspace/catkin_ws/src/my_number_counter/src/oop_number_counter.cpp
#include <ros/ros.h>
#include <std_msgs/Int64.h>
#include <std_srvs/SetBool.h>

// int counter = 0;
// ros::Publisher pub;
class NumberCounter{
    private:
    int counter;
    ros::Publisher pub;
    ros::Subscriber sub;
    ros::ServiceServer server;

    public:
    NumberCounter(ros::NodeHandle *nh){
        counter = 0;
        pub = nh->advertise<std_msgs::Int64>("/number_count", 10);
	    sub = nh->subscribe("/number", 1000, &NumberCounter::callback_receive_number_data, this);

        // Create Server
        server = nh->advertiseService("/reset_number_count", &NumberCounter::reset_number_count, this);

    }

    void callback_receive_number_data(const std_msgs::Int64& msg)
    {
        ROS_INFO("Message Received : %li", msg.data);
        counter += msg.data;
        std_msgs::Int64 new_msg;
        new_msg.data = counter;
        pub.publish(new_msg);

    }

    bool reset_number_count(std_srvs::SetBool::Request &req, std_srvs::SetBool::Response &res){
        if (req.data){
            ROS_INFO("Reset Succeed");
            counter = 0;
            res.success = true;
            res.message = "Counter Reset Success";
        }
        else{
            res.success = false;
            res.message = "Counter Reset False";
        }
        return true;
    }
};



int main(int argc, char **argv)
{
	ros::init(argc, argv, "number_counter");

	ros::NodeHandle nh;
    
    NumberCounter nc = NumberCounter(&nh);
    
	ros::spin(); //Keep node running till we shutdown
}
